File: app/src/test/java/com/mohsenoid/rickandmorty/test/EpisodeDataFactory.java
package com.mohsenoid.rickandmorty.test;


import com.mohsenoid.rickandmorty.data.db.dto.DbEpisodeModel;
import com.mohsenoid.rickandmorty.data.network.dto.NetworkEpisodeModel;
import com.mohsenoid.rickandmorty.domain.entity.EpisodeEntity;

import java.util.ArrayList;
import java.util.List;

public class EpisodeDataFactory {

    private EpisodeDataFactory() { /* this will prevent making a new object */ }

    public static class Db {
        public static DbEpisodeModel makeDbEpisodeModelWithId(int episodeId) {
            return new DbEpisodeModel(
                    episodeId,
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    DataFactory.randomString() + "," + DataFactory.randomString() + "," + DataFactory.randomString(),
                    DataFactory.randomString(),
                    DataFactory.randomString()
            );
        }

        public static DbEpisodeModel makeDbEpisodeModel() {
            return makeDbEpisodeModelWithId(DataFactory.randomInt());
        }

        public static List<DbEpisodeModel> makeDbEpisodesModelList(int count) {
            List<DbEpisodeModel> episodes = new ArrayList<>();

            for (int i = 0; i < count; i++) {
                DbEpisodeModel episode = makeDbEpisodeModel();
                episodes.add(episode);
            }

            return episodes;
        }
    }

    public static class Network {

        public static NetworkEpisodeModel makeNetworkEpisodeModelWithId(int episodeId) {
            return new NetworkEpisodeModel(
                    episodeId,
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    DataFactory.randomStringList(5),
                    DataFactory.randomString(),
                    DataFactory.randomString()
            );
        }

        public static NetworkEpisodeModel makeNetworkEpisodeModel() {
            return makeNetworkEpisodeModelWithId(DataFactory.randomInt());
        }

        public static List<NetworkEpisodeModel> makeNetworkEpisodesModelList(int count) {
            List<NetworkEpisodeModel> episodes = new ArrayList<>();

            for (int i = 0; i < count; i++) {
                NetworkEpisodeModel episode = makeNetworkEpisodeModel();
                episodes.add(episode);
            }

            return episodes;
        }
    }

    public static class Entity {

        public static EpisodeEntity makeEpisodeEntityWithId(int episodeId) {
            return new EpisodeEntity(
                    episodeId,
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    DataFactory.randomIntList(5),
                    DataFactory.randomString(),
                    DataFactory.randomString()
            );
        }

        public static EpisodeEntity makeEpisodeEntity() {
            return makeEpisodeEntityWithId(DataFactory.randomInt());
        }

        public static List<EpisodeEntity> makeEpisodesEntityList(int count) {
            List<EpisodeEntity> episodes = new ArrayList<>();

            for (int i = 0; i < count; i++) {
                EpisodeEntity episode = makeEpisodeEntity();
                episodes.add(episode);
            }

            return episodes;
        }
    }
}

File: app/src/test/kotlin/com/mohsenoid/rickandmorty/test/EpisodeDataFactory.kt
package com.mohsenoid.rickandmorty.test

import com.mohsenoid.rickandmorty.data.api.model.ApiEpisode
import com.mohsenoid.rickandmorty.data.db.model.DbEpisode
import com.mohsenoid.rickandmorty.domain.model.ModelEpisode

object EpisodeDataFactory {

    internal fun makeDbEpisode(episodeId: Int = DataFactory.randomInt()): DbEpisode =
        DbEpisode(
            id = episodeId,
            name = DataFactory.randomString(),
            airDate = DataFactory.randomString(),
            episode = DataFactory.randomString(),
            characterIds = DataFactory.randomIntList(count = 5),
            url = DataFactory.randomString(),
            created = DataFactory.randomString(),
        )

    internal fun makeDbEpisodes(count: Int): List<DbEpisode> {
        val entityEpisodes: MutableList<DbEpisode> = ArrayList()
        repeat(count) {
            val entityEpisode: DbEpisode = makeDbEpisode()
            entityEpisodes.add(entityEpisode)
        }
        return entityEpisodes
    }

    internal fun makeApiEpisode(episodeId: Int = DataFactory.randomInt()): ApiEpisode =
        ApiEpisode(
            id = episodeId,
            name = DataFactory.randomString(),
            airDate = DataFactory.randomString(),
            episode = DataFactory.randomString(),
            characters = DataFactory.randomIntList(count = 5)
                .map { "${DataFactory.randomString()}/$it" },
            url = DataFactory.randomString(),
            created = DataFactory.randomString(),
        )

    internal fun makeApiEpisodes(count: Int): List<ApiEpisode> {
        val episodes: MutableList<ApiEpisode> = ArrayList()
        repeat(count) {
            val episode: ApiEpisode = makeApiEpisode()
            episodes.add(episode)
        }
        return episodes
    }

    internal fun makeEpisode(episodeId: Int = DataFactory.randomInt()): ModelEpisode =
        ModelEpisode(
            id = episodeId,
            name = DataFactory.randomString(),
            airDate = DataFactory.randomString(),
            episode = DataFactory.randomString(),
            characterIds = DataFactory.randomIntList(count = 5),
            url = DataFactory.randomString(),
            created = DataFactory.randomString(),
        )

    internal fun makeEpisodes(count: Int): List<ModelEpisode> {
        val episodes: MutableList<ModelEpisode> = ArrayList()
        repeat(count) {
            val episode: ModelEpisode = makeEpisode()
            episodes.add(episode)
        }
        return episodes
    }
}
